core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -0,0 +1,675 @@
+   */
+  @Override
+  public void setMessageCompression(boolean enable) {
+    synchronized (lock) {
+      if (!committed) {
+        sharedBuffer.add(new MessageCompressionEntry(enable));
+      }
+      for (Substream substream : substreams) {
+        substream.setMessageCompression(enable);
+      }
+    }
+  }
+
+  /** Half-closes if committed, or buffers it first and then half-closes the current substream. */
+  @Override
+  public void halfClose() {
+    synchronized (lock) {
+      if (!committed) {
+        sharedBuffer.add(new HalfCloseEntry());
+      }
+      for (Substream substream : substreams) {
+        substream.halfClose();
+      }
+    }
+  }
+
+  @Override
+  public void setAuthority(String authority) {
+    synchronized (lock) {
+      if (!committed) {
+        sharedBuffer.add(new AuthorityEntry(authority));
+      }
+      for (Substream substream : substreams) {
+        substream.setAuthority(authority);
+      }
+    }
+  }
+
+  @Override
+  public void setDecompressorRegistry(DecompressorRegistry decompressorRegistry) {
+    synchronized (lock) {
+      if (!committed) {
+        sharedBuffer.add(new DecompressorRegistryEntry(decompressorRegistry));
+      }
+      for (Substream substream : substreams) {
+        substream.setDecompressorRegistry(decompressorRegistry);
+      }
+    }
+  }
+
+  /**
+   * Sets the max inbound message size if committed, or buffers it first and then sets it on the
+   * current substream.
+   */
+  @Override
+  public void setMaxInboundMessageSize(int maxSize) {
+    synchronized (lock) {
+      if (!committed) {
+        sharedBuffer.add(new MaxInboundMessageSizeEntry(maxSize));
+      }
+      for (Substream substream : substreams) {
+        substream.setMaxInboundMessageSize(maxSize);
+      }
+    }
+  }
+
+  /**
+   * Sets the max outbound message size if committed, or buffers it first and then sets it on the
+   * current substream.
+   */
+  @Override
+  public void setMaxOutboundMessageSize(int maxSize) {
+    synchronized (lock) {
+      if (!committed) {
+        sharedBuffer.add(new MaxOutboundMessageSizeEntry(maxSize));
+      }
+      for (Substream substream : substreams) {
+        substream.setMaxOutboundMessageSize(maxSize);
+      }
+    }
+  }
+
+  @Override
+  public Attributes getAttributes() {
+    // TODO(zdapeng): Hedging case need be handled properly.
+    if (substreams.size() == 1) {
+      return substreams.iterator().next().getAttributes();
+    }
+    return Attributes.EMPTY;
+  }
+
+  // TODO(zdapeng): implement retry policy.
+  // Retry policy is obtained from the combination of the name resolver plus channel builder, and
+  // passed all the way down to this class.
+  private boolean shouldRetry() {
+    return false;
+  }
+
+  private interface BufferEntry {
+    /** Replays the buffer entry with the given stream. */
+    void runWith(ClientStream substream);
+  }
+
+  private static final class RequestEntry implements BufferEntry {
```

Comment:
:-/ This is a lot of boilerplate and splits the two copies of the method call. If you're doing it just for giving names to these classes, you can still do that in-line in each method. Defining the class in-line (and not anonymous) in the method still allows you to capture final fields.